File: DFMStabilizerCLI.pas
unit DFMStabilizerCLI;

{
  Command-line interface for the DFM stabilizer tool.

  Argument syntax:
    DFMStabilizerTool [-s] [-a:<codepage>] <file|pattern|@listfile> [...]

    -s          Recurse into subdirectories when expanding wildcard patterns.
    -a:<cp>     Treat text DFMs with no UTF-8 BOM as being encoded in the
                given ANSI code page (e.g. -a:1250) instead of assuming
                UTF-8 without BOM. Use this for legacy DFMs that fail with
                "No mapping for the Unicode character exists in the target
                multi-byte code page".
    file        Exact path to a DFM file.
    pattern     Wildcard pattern: *.dfm, path\*.dfm, etc.
    @listfile   Text file listing one path or pattern per line.
                Lines starting with # are treated as comments.

  Exit code: 0 if all files were converted successfully, 1 if any failed.
}

interface

procedure Run;

implementation

uses
  System.SysUtils,
  System.Classes,
  DFMTextStabilizerCore;

// ---------------------------------------------------------------------------

type
  TDFMProcessor = class
  private
    FRecursive   : Boolean;
    FAnsiCodePage: Integer;
    FSuccessCount: Integer;
    FFailCount   : Integer;
    procedure ProcessFile(const AFileName: string);
    procedure ExpandAndProcess(const Pattern: string);
    procedure ProcessListFile(const AListFileName: string);
  public
    constructor Create(ARecursive: Boolean; AAnsiCodePage: Integer);
    procedure ProcessArg(const Arg: string);
    property SuccessCount: Integer read FSuccessCount;
    property FailCount   : Integer read FFailCount;
  end;

constructor TDFMProcessor.Create(ARecursive: Boolean; AAnsiCodePage: Integer);
begin
  inherited Create;
  FRecursive    := ARecursive;
  FAnsiCodePage := AAnsiCodePage;
  FSuccessCount := 0;
  FFailCount    := 0;
end;

procedure TDFMProcessor.ProcessFile(const AFileName: string);
begin
  try
    Write('  ', AFileName, ' ... ');
    if ConvertDFMFile(AFileName, FAnsiCodePage) then
      Writeln('converted')
    else
      Writeln('already up to date');
    Inc(FSuccessCount);
  except
    on E: Exception do
    begin
      Writeln('FAILED: ', E.Message);
      Inc(FFailCount);
    end;
  end;
end;

// Expand a pattern (which may contain * or ?) into actual files, then process
// each one.  If FRecursive is True, descend into subdirectories as well.
procedure TDFMProcessor.ExpandAndProcess(const Pattern: string);
var
  Dir    : string;
  FilePat: string;
  SR     : TSearchRec;
begin
  Dir     := ExtractFilePath(Pattern);
  FilePat := ExtractFileName(Pattern);
  if Dir = '' then
    Dir := '.';

  // Files matching the pattern in this directory
  if FindFirst(IncludeTrailingPathDelimiter(Dir) + FilePat, faAnyFile, SR) = 0 then
  begin
    try
      repeat
        if (SR.Attr and faDirectory) = 0 then
          ProcessFile(IncludeTrailingPathDelimiter(Dir) + SR.Name);
      until FindNext(SR) <> 0;
    finally
      FindClose(SR);
    end;
  end;

  // Recurse into subdirectories
  if FRecursive then
  begin
    if FindFirst(IncludeTrailingPathDelimiter(Dir) + '*', faDirectory, SR) = 0 then
    begin
      try
        repeat
          if ((SR.Attr and faDirectory) <> 0) and
             (SR.Name <> '.') and (SR.Name <> '..') then
            ExpandAndProcess(
              IncludeTrailingPathDelimiter(Dir) + SR.Name +
              PathDelim + FilePat);
        until FindNext(SR) <> 0;
      finally
        FindClose(SR);
      end;
    end;
  end;
end;

procedure TDFMProcessor.ProcessListFile(const AListFileName: string);
var
  List: TStringList;
  I   : Integer;
  Line: string;
begin
  if not FileExists(AListFileName) then
    raise EFileNotFoundException.CreateFmt('List file not found: %s', [AListFileName]);

  List := TStringList.Create;
  try
    List.LoadFromFile(AListFileName);
    for I := 0 to List.Count - 1 do
    begin
      Line := Trim(List[I]);
      if (Line <> '') and not Line.StartsWith('#') then
        ProcessArg(Line);
    end;
  finally
    List.Free;
  end;
end;

procedure TDFMProcessor.ProcessArg(const Arg: string);
begin
  if Arg.StartsWith('@') then
    ProcessListFile(Arg.Substring(1))
  else if (Pos('*', Arg) > 0) or (Pos('?', Arg) > 0) then
    ExpandAndProcess(Arg)
  else
    ProcessFile(Arg);
end;

// ---------------------------------------------------------------------------

const
  AnsiCodePageSwitch = '-a:';

function IsAnsiCodePageArg(const Arg: string): Boolean;
begin
  Result := Arg.StartsWith(AnsiCodePageSwitch, True);
end;

// Extract the numeric code page from an "-a:<codepage>" argument.
// Accepts a bare number (-a:1250) or a name containing one (-a:CP1250).
function ExtractAnsiCodePage(const Arg: string): Integer;
var
  Value : string;
  Digits: string;
  C     : Char;
begin
  Value := Arg.Substring(Length(AnsiCodePageSwitch));
  Digits := '';
  for C in Value do
    if CharInSet(C, ['0'..'9']) then
      Digits := Digits + C;
  if not TryStrToInt(Digits, Result) then
    raise Exception.CreateFmt('Invalid ANSI code page: "%s"', [Arg]);
end;

procedure PrintUsage;
begin
  Writeln('Usage: DFMStabilizerTool [-s] [-a:<codepage>] <file|pattern|@listfile> [...]');
  Writeln;
  Writeln('Converts DFM files in-place to the stabilized UTF-8 text format:');
  Writeln('  - strings are not broken at 64 characters (limit raised to 700)');
  Writeln('  - embedded newlines (#13/#10) cause a line break at that position');
  Writeln('  - non-ASCII characters are written literally as UTF-8');
  Writeln('  - file always starts with a UTF-8 BOM');
  Writeln;
  Writeln('Options:');
  Writeln('  -s          Recurse into subdirectories when expanding wildcard patterns');
  Writeln('  -a:<cp>     Decode text DFMs with no UTF-8 BOM using ANSI code page <cp>');
  Writeln('              (e.g. -a:1250 for Central European / Czech) instead of');
  Writeln('              assuming UTF-8 without BOM. Use this if conversion fails with');
  Writeln('              "No mapping for the Unicode character exists in the target');
  Writeln('              multi-byte code page".');
  Writeln;
  Writeln('Arguments:');
  Writeln('  file        Exact path to a DFM file');
  Writeln('  pattern     Wildcard pattern  (e.g.  *.dfm   or   forms\*.dfm)');
  Writeln('  @listfile   Text file with one path/pattern per line (# = comment)');
  Writeln;
  Writeln('Examples:');
  Writeln('  DFMStabilizerTool MainForm.dfm');
  Writeln('  DFMStabilizerTool -s *.dfm');
  Writeln('  DFMStabilizerTool -s src\*.dfm @extra_forms.txt');
  Writeln('  DFMStabilizerTool @all_forms.txt');
  Writeln('  DFMStabilizerTool -a:1250 legacy\*.dfm');
end;

// ---------------------------------------------------------------------------

procedure Run;
var
  Recursive   : Boolean;
  AnsiCodePage: Integer;
  Processor   : TDFMProcessor;
  I           : Integer;
  Arg         : string;
begin
  if ParamCount = 0 then
  begin
    PrintUsage;
    Halt(1);
  end;

  Recursive    := False;
  AnsiCodePage := 0;
  for I := 1 to ParamCount do
  begin
    Arg := ParamStr(I);
    if SameText(Arg, '-s') then
      Recursive := True
    else if IsAnsiCodePageArg(Arg) then
      AnsiCodePage := ExtractAnsiCodePage(Arg);
  end;

  Processor := TDFMProcessor.Create(Recursive, AnsiCodePage);
  try
    for I := 1 to ParamCount do
    begin
      Arg := ParamStr(I);
      if SameText(Arg, '-s') or IsAnsiCodePageArg(Arg) then
        Continue;
      Processor.ProcessArg(Arg);
    end;

    Writeln;
    if Processor.SuccessCount + Processor.FailCount = 0 then
      Writeln('No files matched.')
    else
      Writeln(Format('%d converted, %d failed.',
        [Processor.SuccessCount, Processor.FailCount]));

    if Processor.FailCount > 0 then
      Halt(1);
  finally
    Processor.Free;
  end;
end;

end.

File: dfm_stabilizer_cli.py
"""
Command-line interface for the Python port of DFMStabilizerTool.

Argument syntax:
  dfm_stabilizer_tool.py [-s] [-a:<codepage>] <file|pattern|@listfile> [...]

  -s          Recurse into subdirectories when expanding wildcard patterns.
  -a:<cp>     Decode text DFMs with no UTF-8 BOM using ANSI code page <cp>
              (e.g. -a:1250 for Central European / Czech) instead of
              assuming UTF-8 without BOM. Use this if conversion fails with
              "No mapping for the Unicode character exists in the target
              multi-byte code page" style errors.
  file        Exact path to a DFM file.
  pattern     Wildcard pattern: *.dfm, path\\*.dfm, etc.
  @listfile   Text file listing one path or pattern per line.
              Lines starting with # are treated as comments.

Only text DFMs are supported (UTF-8 with/without BOM, or a single-byte ANSI
code page). Legacy binary DFMs (signature 0xFF 0x0A) are reported as failed;
use the original Delphi DFMStabilizerTool for those.

Exit code: 0 if all files were converted successfully, 1 if any failed.
"""

from __future__ import annotations

import fnmatch
import os
import re
import sys
from pathlib import Path
from typing import List, Optional

from dfm_text_stabilizer_core import DFMError, stabilize_dfm_bytes

ANSI_CODE_PAGE_SWITCH = "-a:"


def is_ansi_code_page_arg(arg: str) -> bool:
    return arg.lower().startswith(ANSI_CODE_PAGE_SWITCH)


def extract_ansi_code_page(arg: str) -> int:
    value = arg[len(ANSI_CODE_PAGE_SWITCH):]
    digits = "".join(re.findall(r"\d", value))
    if not digits:
        raise DFMError(f'Invalid ANSI code page: "{arg}"')
    return int(digits)


def convert_dfm_file(file_name: str, ansi_code_page: Optional[int] = None) -> bool:
    """
    Convert file_name in-place to the stabilized text format.
    Returns True if the file was rewritten, False if already up to date.
    Raises DFMError (or an OSError) on failure.
    """
    with open(file_name, "rb") as f:
        original = f.read()

    stabilized = stabilize_dfm_bytes(original, ansi_code_page)

    if stabilized == original:
        return False

    tmp_name = file_name + ".dfmstab.tmp"
    try:
        with open(tmp_name, "wb") as f:
            f.write(stabilized)
        os.replace(tmp_name, file_name)
    except Exception:
        if os.path.exists(tmp_name):
            os.remove(tmp_name)
        raise

    return True


class DFMProcessor:
    def __init__(self, recursive: bool, ansi_code_page: Optional[int]):
        self.recursive = recursive
        self.ansi_code_page = ansi_code_page
        self.success_count = 0
        self.fail_count = 0

    def process_file(self, file_name: str) -> None:
        try:
            print(f"  {file_name} ... ", end="")
            if convert_dfm_file(file_name, self.ansi_code_page):
                print("converted")
            else:
                print("already up to date")
            self.success_count += 1
        except Exception as exc:  # noqa: BLE001 - mirror the Delphi try/except-per-file behavior
            print(f"FAILED: {exc}")
            self.fail_count += 1

    def expand_and_process(self, pattern: str) -> None:
        directory = os.path.dirname(pattern) or "."
        file_pattern = os.path.basename(pattern)

        if os.path.isdir(directory):
            try:
                entries = sorted(os.listdir(directory))
            except OSError:
                entries = []
            for entry in entries:
                full_path = os.path.join(directory, entry)
                if os.path.isfile(full_path) and fnmatch.fnmatch(entry, file_pattern):
                    self.process_file(full_path)

            if self.recursive:
                for entry in entries:
                    full_path = os.path.join(directory, entry)
                    if os.path.isdir(full_path):
                        self.expand_and_process(os.path.join(full_path, file_pattern))

    def process_list_file(self, list_file_name: str) -> None:
        if not os.path.isfile(list_file_name):
            raise FileNotFoundError(f"List file not found: {list_file_name}")

        with open(list_file_name, "r", encoding="utf-8-sig") as f:
            lines = f.readlines()

        for line in lines:
            line = line.strip()
            if line and not line.startswith("#"):
                self.process_arg(line)

    def process_arg(self, arg: str) -> None:
        if arg.startswith("@"):
            self.process_list_file(arg[1:])
        elif "*" in arg or "?" in arg:
            self.expand_and_process(arg)
        else:
            self.process_file(arg)


def print_usage() -> None:
    print("Usage: dfm_stabilizer_tool.py [-s] [-a:<codepage>] <file|pattern|@listfile> [...]")
    print()
    print("Converts DFM files in-place to the stabilized UTF-8 text format:")
    print("  - strings are not broken at 64 characters (limit raised to 700)")
    print("  - embedded newlines (#13/#10) cause a line break at that position")
    print("  - non-ASCII characters are written literally as UTF-8")
    print("  - file always starts with a UTF-8 BOM")
    print()
    print("Only text DFMs are supported (legacy binary DFMs are reported as failed;")
    print("use the Delphi DFMStabilizerTool.exe for those).")
    print()
    print("Options:")
    print("  -s          Recurse into subdirectories when expanding wildcard patterns")
    print("  -a:<cp>     Decode text DFMs with no UTF-8 BOM using ANSI code page <cp>")
    print("              (e.g. -a:1250 for Central European / Czech) instead of")
    print("              assuming UTF-8 without BOM. Use this if conversion fails with")
    print('              "No mapping for the Unicode character exists in the target')
    print('              multi-byte code page".')
    print()
    print("Arguments:")
    print("  file        Exact path to a DFM file")
    print("  pattern     Wildcard pattern  (e.g.  *.dfm   or   forms\\*.dfm)")
    print("  @listfile   Text file with one path/pattern per line (# = comment)")
    print()
    print("Examples:")
    print("  dfm_stabilizer_tool.py MainForm.dfm")
    print("  dfm_stabilizer_tool.py -s *.dfm")
    print("  dfm_stabilizer_tool.py -s src\\*.dfm @extra_forms.txt")
    print("  dfm_stabilizer_tool.py @all_forms.txt")
    print("  dfm_stabilizer_tool.py -a:1250 legacy\\*.dfm")


def run(argv: Optional[List[str]] = None) -> int:
    args = sys.argv[1:] if argv is None else argv

    if not args:
        print_usage()
        return 1

    recursive = False
    ansi_code_page: Optional[int] = None
    for arg in args:
        if arg == "-s":
            recursive = True
        elif is_ansi_code_page_arg(arg):
            try:
                ansi_code_page = extract_ansi_code_page(arg)
            except DFMError as exc:
                print(f"Fatal: {exc}", file=sys.stderr)
                return 1

    processor = DFMProcessor(recursive, ansi_code_page)
    for arg in args:
        if arg == "-s" or is_ansi_code_page_arg(arg):
            continue
        processor.process_arg(arg)

    print()
    total = processor.success_count + processor.fail_count
    if total == 0:
        print("No files matched.")
    else:
        print(f"{processor.success_count} converted, {processor.fail_count} failed.")

    return 1 if processor.fail_count > 0 else 0
